src/Controllers/LoginController.php

```diff
@@ -28,17 +28,25 @@ public function loginView()
     /**
      * Attempts to log the user in.
      *
-     * @return Response|string
+     * @return RedirectResponse
      */
-    public function loginAction()
+    public function loginAction(): RedirectResponse
     {
-        /** @var IncomingRequest $request */
-        $request = service('request');
+        // Validate here first, since some things,
+        // like the password, can only be validated properly here.
+        $rules = $this->getValidationRules();
 
-        $credentials             = $request->getPost(setting('Auth.validFields'));
+        /** @var Validation $validation */
```

Comment:
See the error.
> PHPDoc tag @var for variable $validation contains unknown class CodeIgniter\Shield\Controllers\Validation.

However the `$validation` below is also not needed, because the controller will have `$this->validator`.